contracts/src/libraries/PlonkVerifier.sol

```diff
@@ -134,37 +113,57 @@ library PlonkVerifier {
+    function _verify(
         IPlonkVerifier.VerifyingKey memory verifyingKey,
         uint256[8] memory publicInput,
         IPlonkVerifier.PlonkProof memory proof
-    ) internal view returns (PcsInfo memory res) {
-        if (publicInput.length != 8) revert WrongPlonkVK();
+    ) private view returns (bool) {
+        if (verifyingKey.numInputs != 8) revert WrongPlonkVK();
 
         Challenges memory chal = _computeChallenges(verifyingKey, publicInput, proof);
 
         Poly.EvalDomain memory domain = Poly.newEvalDomain(verifyingKey.domainSize);
         // pre-compute evaluation data
         Poly.EvalData memory evalData = Poly.evalDataGen(domain, chal.zeta, publicInput);
 
-        // compute opening proof in poly comm.
-        uint256[] memory commScalars = new uint256[](30);
-        BN254.G1Point[] memory commBases = new BN254.G1Point[](30);
+        // in the final pairing check: e(a, [x]_2) =?= e(b, [1]_2)
+        BN254.G1Point memory a;
+        BN254.G1Point memory b;
+
+        // a = openingProof + shiftedOpeningProof^u
+        // in Plonk paper: "[Wz]1 + u · [Wzω]1"
+        a = BN254.add(proof.zeta, BN254.scalarMul(proof.zetaOmega, BN254.ScalarField.wrap(chal.u)));
 
-        uint256 eval =
-            _prepareOpeningProof(verifyingKey, evalData, proof, chal, commScalars, commBases);
+        // computing b in Plonk paper: "z · [Wz]1 + uzω · [Wzω]1 + [F]1 − [E]1"
+        (BN254.G1Point memory e1, BN254.G1Point memory f1) =
+            _preparePolyCommitments(verifyingKey, chal, evalData, proof);
+        b = BN254.add(f1, BN254.negate(e1)); // [F]1 − [E]1
+        // b += proof.zeta^chal.zeta or "z · [Wz]1"
+        b = BN254.add(b, BN254.scalarMul(proof.zeta, BN254.ScalarField.wrap(chal.zeta)));
 
-        uint256 zeta = chal.zeta;
-        uint256 omega = domain.elements[1]; // groupGen
         uint256 p = BN254.R_MOD;
-        uint256 zetaOmega;
+        uint256 scalar;
         assembly {
-            zetaOmega := mulmod(zeta, omega, p)
+            // chal.zeta
+            scalar := mload(add(chal, 0xa0))
+            // chal.zeta * groupGen or nextEvalPoint or zetaOmega
+            scalar := mulmod(scalar, mload(add(mload(add(domain, 0x40)), 0x20)), p)
+            // u * zetaOmega or "uzω"
+            scalar := mulmod(scalar, mload(add(chal, 0xe0)), p)
         }
+        // b += proof.zetaOmega^(u * chal.zeta * groupGen)
+        b = BN254.add(b, BN254.scalarMul(proof.zetaOmega, BN254.ScalarField.wrap(scalar)));
 
-        res = PcsInfo(
-            chal.u, zeta, zetaOmega, eval, commScalars, commBases, proof.zeta, proof.zetaOmega
-        );
+        BN254.G2Point memory betaH = BN254.G2Point({
+            x0: BN254.BaseField.wrap(BETA_H_X1),
+            x1: BN254.BaseField.wrap(BETA_H_X0),
+            y0: BN254.BaseField.wrap(BETA_H_Y1),
+            y1: BN254.BaseField.wrap(BETA_H_Y0)
+        });
+        // Check e(A, [x]2) ?= e(B, [1]2)
```

Comment:
```suggestion
        // Check e(A, [x]2) =?= e(B, [1]2)
```

Reply:
done in [025cd8a](https://github.com/EspressoSystems/espresso-sequencer/pull/1940/commits/025cd8aae06b9fd9771bfce047517560626aa31f)